### PR TITLE
feat: discover slsa v1 provenances for npm packages

### LINKED ISSUE #549: Discover SLSA v1 provenances for npm packages
`npm` is producing SLSA v1 provenances, e.g., see [sigstore@2.1.0](https://registry.npmjs.org/-/npm/v1/attestations/sigstore@2.1.0). We need to support these provenances.

### PR DESCRIPTION
**Pull request by benmss**
This PR adds:

- Support for downloading SLSA v1 provenance files for NPM packages
- Support for v1 in-toto attestation files, including content validation

It also removes comments relating to intended use of the in-toto attestation library after the investigation reported in #426 

Closes #549

### CODE REVIEW

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 1]
```diff
@@ -1,9 +1,37 @@
...
-from typing import TypedDict
+from __future__ import annotations
+
+from collections.abc import Callable
+from typing import TypedDict, TypeGuard
+
+from macaron.slsa_analyzer.provenance.intoto.errors import ValidateInTotoPayloadError
+from macaron.util import JsonType
+
+VALID_ALGORITHMS = [
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:14]: Since this is only used in the `is_valid_digest_set` function, and the link to the documentation mentioning these algorithms is also there, do you think it is better to move this list into that function, for closer proximity to the documentation link? Either that or we may need to include the documentation link here because just by looking at this list, I feel it is a bit hard to tell where it came from and what it is for.

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 2]
```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
...
+        raise ValidateInTotoPayloadError("The subject name is not valid.")
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_valid = _validate_property(subject, "uri", True, lambda x: isinstance(x, str))
+    content_valid = _validate_property(subject, "content", True, lambda x: isinstance(x, str))
+    digest_valid = _validate_property(subject, "digest", True, is_valid_digest_set)
+    if not (uri_valid or content_valid or digest_valid):
+        raise ValidateInTotoPayloadError(
+            "One of 'uri', 'digest', or 'content' must be present and valid within 'subject'."
+        )
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:166]: Combining different issues into a single error message may not be the best idea since it might be hard to tell what the exact issue is when one comes up.

Since you pass the name of the attribute into the function, it may be better to raise the error within the `_validate_property` function itself with a more specific context and error message, rather than returning a `bool` value, only just to convert it to an exception later. All `if` statements like this one can be removed in that case.

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 3]
```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
...
+) -> bool:
+    """Validate the existence and type of target within the passed Json object."""
+    if not isinstance(object_, dict):
+        return False
+
+    target = object_.get(target_name)
+    if not target:
+        return not required
+
+    return validator_function(target)
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:237]: Nit: `property_name` (or `key`) and `value` would be a more familiar naming scheme for me than `target_name` and `target`. Also in that case, `validator_function` can be called `value_validator`, since it is clearer what is being validated, and the word `function` may not be as necessary since it is already type-annotated as a `Callable`. What do you think?

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 4]
```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
...
+    Parameters
+    ----------
+    subject : JsonType
+        The JSON element representing a single subject.
+
+    Returns
+    -------
+    TypeGuard[InTotoSubject]
+        ``True`` if the subject element is valid, in which case its type is narrowed to an
+        ``InTotoSubject``; ``False`` otherwise.
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:143]:
```suggestion
    Returns
    -------
    TypeGuard[InTotoV1ResourceDescriptor]
        ``True`` if the subject element is valid, in which case its type is narrowed to an
        ``InTotoV1ResourceDescriptor``; ``False`` otherwise.
```

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 5]
```diff
@@ -12,3 +44,211 @@ class InTotoV1Statement(TypedDict):
...
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_state = _validate_property(subject, "uri", True, lambda x: isinstance(x, str), False)
+    content_state = _validate_property(subject, "content", True, lambda x: isinstance(x, str), False)
+    digest_state = _validate_property(subject, "digest", True, is_valid_digest_set, False)
+    if _ValidityState.VALID not in (uri_state, content_state, digest_state):
+        raise ValidateInTotoPayloadError(
+            f"One of 'uri', 'digest', or 'content' must be present and valid within 'subject'. URI: {uri_state.name}. "
+            f"Content: {content_state.name}. DIGEST: {digest_state.name}."
+        )
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:167]: I guess something else you can do instead of using this `report_errors` param, is to return from `_validate_property` the object you are validating (if it is not there, return `None`). Then you can use the values to check for the co-existence of attributes.
```suggestion
    # At least one of 'uri', 'digest', and 'content' must be valid and present.
    uri = _validate_property(subject, "uri", required=False, validator=lambda x: isinstance(x, str))
    content = _validate_property(subject, "content", required=False, validator=lambda x: isinstance(x, str))
    digest = _validate_property(subject, "digest", required=False, validator=is_valid_digest_set)
    if not any([uri, content, digest]):
        raise ValidateInTotoPayloadError("One of 'uri', 'digest', or 'content' must be present and valid within 'subject'")
```

The `_validate_property` function then can be like this:

```python
def _validate_property(
    object_: dict[str, JsonType],
    key: str,
    required: bool,
    validator: Callable[[JsonType], bool],
) -> _ValidityState:
    """Validate the existence and type of target within the passed Json object."""
    value = object_.get(key)

    if not value:
        if required:
            raise ValidateInTotoPayloadError(f"The attribute {key} of the in-toto subject is missing.")
        return None

    if not validator(value):
        raise ValidateInTotoPayloadError(f"The attribute {key} of the in-toto subject is invalid.")

    return value
```

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 6]
```diff
@@ -12,3 +44,211 @@ class InTotoV1Statement(TypedDict):
...
+    validator: Callable[[JsonType], bool],
+    report_errors: bool = True,
+) -> _ValidityState:
+    """Validate the existence and type of target within the passed Json object."""
+    if not isinstance(object_, dict):
+        # The passed object is checked before this function is called, but mypy requires it be explicitly checked here
+        #  too. Also known as: this branch should never be taken.
+        return _ValidityState.INVALID
+
+    value = object_.get(key)
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:233]: Since you have already checked if `subject` is a `dict`, `object_` here can be type-annotated as `dict`. That way you don't need the `if` statement to narrow the type again to make `mypy` happy.

```suggestion
def _validate_property(
    object_: dict[str, JsonType],
    key: str,
    required: bool,
    validator: Callable[[JsonType], bool],
    report_errors: bool = True,
) -> _ValidityState:
    """Validate the existence and type of target within the passed Json object."""
   value = object_.get(key)
```

File: src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py [hunk 7]
```diff
@@ -12,3 +43,189 @@ class InTotoV1Statement(TypedDict):
...
+    digest = _validate_property(subject, "digest", is_valid_digest_set)
+    if not any([uri, content, digest]):
+        raise ValidateInTotoPayloadError(
+            "One of 'uri', 'digest', or 'content' must be present and valid within 'subject'."
+        )
+
+    _validate_property(subject, "name", lambda x: isinstance(x, str))
+    _validate_property(subject, "downloadLocation", lambda x: isinstance(x, str))
+    _validate_property(subject, "mediaType", lambda x: isinstance(x, str))
+    _validate_property(subject, "annotations", is_valid_annotation_map)
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py:170]: Wouldn't a lambda be enough in this case, since you only need to check with `isinstance` and this is consistent with the rest of the function?

```suggestion
    _validate_property(subject, "annotations", lambda x: isinstance(x, dict))
```

File: src/macaron/slsa_analyzer/provenance/intoto/v01/__init__.py [hunk 8]
```diff
@@ -39,10 +39,6 @@ def validate_intoto_statement(payload: dict[str, JsonType]) -> TypeGuard[InTotoV
 
     Specification: https://github.com/in-toto/attestation/tree/main/spec/v0.1.0#statement.
 
-    TODO: Consider using the in-toto-attestation package (https://github.com/in-toto/attestation/tree/main/python),
```
Comment (nathanwn) [inline src/macaron/slsa_analyzer/provenance/intoto/v01/__init__.py:42]: I have a question about this: I vaguely remember that you mentioned looking at this package and seeing some issues with it. I wonder what was the issue you saw and whether we want to reconsider using this package again in the future.

Comment (benmss) [reply src/macaron/slsa_analyzer/provenance/intoto/v01/__init__.py:42]: It does not provide any methods relating to verifying the validity of a given attestation file in the straightforward way that we require.

File: tests/slsa_analyzer/package_registry/test_npm_registry.py [hunk 9]
```diff
@@ -17,14 +17,14 @@
 
 @pytest.fixture(name="npm_registry")
 def create_npm_registry() -> NPMRegistry:
-    """Create an npm registry instance."""
+    """Create an NPM registry instance."""
```
Comment (behnazh-w) [inline tests/slsa_analyzer/package_registry/test_npm_registry.py:20]: `npm` should be lowercase. See https://docs.npmjs.com/cli/v7/using-npm/registry

Comment (benmss) [reply tests/slsa_analyzer/package_registry/test_npm_registry.py:20]: But is the comment referring to the class or the external system the class represents? I'll change it, I'm just commenting that it is a bit hard to follow their preferred naming convention when it directly conflicts with our own class naming convention and we ignore it there.

Comment (behnazh-w) [reply tests/slsa_analyzer/package_registry/test_npm_registry.py]: The comment is referring to the external `npm` registry. But can I see how it gets tricky to figure out the right naming convention.

**Review by behnazh-w — APPROVED** [review]
